### ISSUE TITLE
verifyDeveloperNodeModules() fails to discover "firebase-admin"

### ISSUE THREAD

**Issue by ExtraBB** [firebase/firebase-tools#1262]
### Related issues
#1243 and #1258

### [REQUIRED] Version info

**node:** 8.16.0

**firebase-functions:** 2.3.0

**firebase-tools:** 6.9.1 (**NOTE:** Works on 6.8.0)

**firebase-admin:** 7.3.0

I am running this on Windows 10.

### [REQUIRED] Test case

A freshly generated firebase project using `firebase init` with functions.

### [REQUIRED] Steps to reproduce

1.  Run `npm run start` (or `firebase functions:shell`)

### [REQUIRED] Expected behavior

The emulator starts correctly

### [REQUIRED] Actual behavior

The following logs are shown:

```bash
!  The Cloud Functions emulator requires the module "firebase-admin" to be installed. This package is in your package.json, but it's not available. You probably need to run "npm install" in your functions directory.
i  functions: Your functions could not be parsed due to an issue with your node_modules (see above)
```
### Were you able to successfully deploy your functions?

Yes, deploying works fine. The deployed function (hello world) as well

**Comment by ExtraBB** [comment]
The emulator works on `firebase-tools@6.8.0`. This issue is likely related to #1243 and #1258

**Comment by yerzhant** [comment]
Experiencing the same issue:

```
✔  functions: Using node@8 from host.
✔  functions: Emulator started at http://localhost:5001
i  functions: Watching "/home/yerzhan/projects/askimam/functions" for Cloud Functions...
⚠  The Cloud Functions emulator requires the module "firebase-admin" to be installed. This package is in your package.json, but it's not available. You probably need to run "npm install" in your functions directory.
i  functions: Your functions could not be parsed due to an issue with your node_modules (see above)
```

```
firebase -V
6.9.1
```

```
  "dependencies": {
    "firebase-admin": "^7.3.0",
    "firebase-functions": "^2.3.0"
  },
```

Though the firebase-admin package does exist in node_modules.

**Comment by yerzhant** [comment]
```
[2019-05-09T09:19:26.091Z] ----------------------------------------------------------------------
[2019-05-09T09:19:26.093Z] Command:       /home/yerzhan/node-v8.12.0-linux-x64/bin/node /home/yerzhan/node-v8.12.0-linux-x64/bin/firebase emulators:start --debug
[2019-05-09T09:19:26.093Z] CLI Version:   6.9.1
[2019-05-09T09:19:26.093Z] Platform:      linux
[2019-05-09T09:19:26.093Z] Node Version:  v8.12.0
[2019-05-09T09:19:26.094Z] Time:          Thu May 09 2019 15:19:26 GMT+0600 (+06)
[2019-05-09T09:19:26.094Z] ----------------------------------------------------------------------

[2019-05-09T09:19:26.099Z] > command requires scopes: ["email","openid","https://www.googleapis.com/auth/cloudplatformprojects.readonly","https://www.googleapis.com/auth/firebase","https://www.googleapis.com/auth/cloud-platform"]
[2019-05-09T09:19:26.099Z] > authorizing via signed-in user
[2019-05-09T09:19:26.100Z] > refreshing access token with scopes: ["email","https://www.googleapis.com/auth/cloud-platform","https://www.googleapis.com/auth/cloudplatformprojects.readonly","https://www.googleapis.com/auth/firebase","openid"]
[2019-05-09T09:19:26.100Z] >>> HTTP REQUEST POST https://www.googleapis.com/oauth2/v3/token  
 <request body omitted>
[2019-05-09T09:19:26.891Z] <<< HTTP RESPONSE 200 content-type=application/json; charset=utf-8, vary=X-Origin, Referer, Origin,Accept-Encoding, date=Thu, 09 May 2019 09:19:26 GMT, server=scaffolding on HTTPServer2, cache-control=private, x-xss-protection=0, x-frame-options=SAMEORIGIN, x-content-type-options=nosniff, alt-svc=quic=":443"; ma=2592000; v="46,44,43,39", accept-ranges=none, transfer-encoding=chunked
[2019-05-09T09:19:26.901Z] >>> HTTP REQUEST GET https://cloudresourcemanager.googleapis.com/v1/projects/azan-kz-ask-imam  
 
[2019-05-09T09:19:28.625Z] <<< HTTP RESPONSE 200 content-type=application/json; charset=UTF-8, vary=X-Origin, Referer, Origin,Accept-Encoding, date=Thu, 09 May 2019 09:19:28 GMT, server=ESF, cache-control=private, x-xss-protection=0, x-frame-options=SAMEORIGIN, x-content-type-options=nosniff, server-timing=gfet4t7; dur=923, alt-svc=quic=":443"; ma=2592000; v="46,44,43,39", accept-ranges=none, transfer-encoding=chunked
i  Starting emulators: ["functions"]
✔  functions: Using node@8 from host.
[2019-05-09T09:19:28.637Z] >>> HTTP REQUEST GET https://mobilesdk-pa.googleapis.com/v1/projects/674391970261:getServerAppConfig  
 
[2019-05-09T09:19:29.356Z] <<< HTTP RESPONSE 200 content-type=application/json; charset=UTF-8, vary=X-Origin, Referer, Origin,Accept-Encoding, date=Thu, 09 May 2019 09:19:29 GMT, server=ESF, cache-control=private, x-xss-protection=0, x-frame-options=SAMEORIGIN, x-content-type-options=nosniff, alt-svc=quic=":443"; ma=2592000; v="46,44,43,39", accept-ranges=none, transfer-encoding=chunked
✔  functions: Emulator started at http://localhost:5001
i  functions: Watching "/home/yerzhan/projects/askimam/functions" for Cloud Functions...
[2019-05-09T09:19:29.813Z] Functions runtime initialized.
[2019-05-09T09:19:29.813Z] Disabled runtime features: undefined
⚠  The Cloud Functions emulator requires the module "firebase-admin" to be installed. This package is in your package.json, but it's not available. You probably need to run "npm install" in your functions directory.
i  functions: Your functions could not be parsed due to an issue with your node_modules (see above)

```

**Comment by githithu** [comment]
https://stackoverflow.com/questions/56056244/im-having-trouble-while-serving-or-deploying-firebase-hosting-functions?noredirect=1#comment98753397_56056244

**Comment by namcoder** [comment]
I have the same and found solution:
You can use the alias command like this: "yarn serve"
It will call from the packages.json , in "scripts" block

**Comment by diogolessa** [comment]
Same issue here! Tried to found a workaround during 3 hours, but without success.
Today I just thought about adding firebase-admin to global npm, but not sure if it should work.
Will try out your suggestion @namcoder

**Comment by ExtraBB** [comment]
@diogolessa Downgrading to firebase-tools 6.8.0 also fixes the problem, is that not an option for you?

**Comment by diogolessa** [comment]
Sure, can be... Looks like a better approach. Though, would be nice to have all the newest versions working normally for beginners like me. Thanks @ExtraBB

**Comment by caputchinefrobles** [comment]
Same here... It happens on linux as well.

**Comment by RafaelH-us** [comment]
I was having this issue I removed firebase-tools and reinstalled firebase-tools@6.8.0.

**Comment by chelopme** [comment]
Same issue.. downgrading is the only way to make it work 

Only setting that currently works
firebase-functions: 2.3.0
firebase-tools: 6.8.0
firebase-admin: 7.3.0

**Comment by abeisgoat** [comment]
Hey folks, sorry about the trouble this issue is causing. For now we recommend downgrading to `6.8.0` while we work on a patch.

**Comment by abeisgoat** [comment]
This issue is caused by the use of the native [`require.resolve`](https://github.com/firebase/firebase-tools/blob/master/src/emulator/functionsEmulatorRuntime.ts#L167) in `verifyDeveloperNodeModules()` where [`slowRequireResolve`](https://github.com/firebase/firebase-tools/blob/master/src/emulator/functionsEmulatorRuntime.ts#L20) should be used. The [other use](https://github.com/firebase/firebase-tools/blob/master/src/emulator/functionsEmulatorRuntime.ts#L595) of native `require.resolve` will also suffer from the same issue.

This was not caught in our tests because `require.resolve` resolved to the dev dependencies of `firebase-tools` when running in tests, so it made it appear that there was no behavior difference between `require-resolve` and `slowRequireResolve`

**Comment by powolnymarcel** [comment]
This is a bug in `firebase-tools` versions `6.9.0` and `6.9.1`.  The solution is to downgrade to `6.8.0` for now:

```
npm install -g firebase-tools@6.8.0
```

ref: https://stackoverflow.com/a/56065094/4319131

**Comment by LVeach** [comment]
Thank you for the updates.  Downgrading to 6.8.0 worked for us

**Comment by ChrisBrownie55** [comment]
This is not currently documented in the [firebase docs for running functions locally](https://firebase.google.com/docs/functions/local-emulator) and definitely needs to be if a fix isn't already ready.

**Comment by abeisgoat** [comment]
Alright folks, we've just merged in a fix for this which will go out in `6.9.2` today. We may wait on a fix for #1265 before releasing, but if that doesn't make it, this will go out as is. Once it's out, I'll close this bug, but please re-open if the issue is still happening for you.

**Comment by samtstern** [comment]
Fixed by #1263

**Comment by samtstern** [comment]
Version `6.9.2` has been released with this fix.

**Comment by ChrisBrownie55** [comment]
Getting the following error, wondering if it's related to this new release:

```
AssertionError [ERR_ASSERTION]: missing path
    at Module.require (module.js:595:3)
    at require (internal/module.js:11:18)
    at InitializeFirebaseAdminStubs
```

**Comment by LVeach** [comment]
upgraded from 6.8.0 to 6.9.2 and am now seeing:
 functions: Using node@8 from host.
+  functions: Emulator started at http://localhost:5001
i  functions: Watching "C:\Users\Luke\Documents\Projects\MenuHelper\gapp\fb\menu-helper-fb\functions" for Cloud Functions...
!  The Cloud Functions emulator requires the module "firebase-functions-test" to be installed as a development dependency. To fix this, run "npm install --save-dev firebase-functions-test" in your functions directory.
i  functions: Your functions could not be parsed due to an issue with your node_modules (see above)

**Comment by naeemshaikh27** [comment]
I am getting a lot of errors after upgrading to 6.9.2. 
1. https://github.com/firebase/firebase-tools/issues/1106
2. https://github.com/firebase/firebase-tools/issues/1265
and few more bugs after trying to tackle down the existing ones.
this upgrade is way more broken. had to downgrade to  6.8.0

**Comment by shaunluttin** [comment]
We are seeing this error in `"firebase-tools": "^6.10.0"`.

**Comment by jaschaio** [comment]
Seeing this error with `firebase-tools: 6.10.0` and node `8.14.0`. Firebase admin is `8.0.0` and firebase functions `6.10.0`:

```
AssertionError [ERR_ASSERTION]: missing path
    at Module.require (module.js:595:3)
    at require (internal/module.js:11:18)
    at InitializeFirebaseAdminStubs (/Users/Jascha/.nvm/versions/node/v8.14.0/lib/node_modules/firebase-tools/lib/emulator/functionsEmulatorRuntime.js:231:18)
    at /Users/Jascha/.nvm/versions/node/v8.14.0/lib/node_modules/firebase-tools/lib/emulator/functionsEmulatorRuntime.js:451:9
    at Generator.next (<anonymous>)
    at /Users/Jascha/.nvm/versions/node/v8.14.0/lib/node_modules/firebase-tools/lib/emulator/functionsEmulatorRuntime.js:7:71
    at new Promise (<anonymous>)
    at __awaiter (/Users/Jascha/.nvm/versions/node/v8.14.0/lib/node_modules/firebase-tools/lib/emulator/functionsEmulatorRuntime.js:3:12)
    at main (/Users/Jascha/.nvm/versions/node/v8.14.0/lib/node_modules/firebase-tools/lib/emulator/functionsEmulatorRuntime.js:421:12)
    at Object.<anonymous> (/Users/Jascha/.nvm/versions/node/v8.14.0/lib/node_modules/firebase-tools/lib/emulator/functionsEmulatorRuntime.js:511:5)
```

If I change node to version `10.15.3` (and as well udpate the `engine` to `10` within the package.json) the error changes to this one:

```
TypeError [ERR_INVALID_ARG_VALUE]: The argument 'id' must be a non-empty string. Received ''
    at Module.require (internal/modules/cjs/loader.js:634:11)
    at require (internal/modules/cjs/helpers.js:22:18)
    at InitializeFirebaseAdminStubs (/Users/Jascha/.nvm/versions/node/v10.15.3/lib/node_modules/firebase-tools/lib/emulator/functionsEmulatorRuntime.js:231:18)
    at /Users/Jascha/.nvm/versions/node/v10.15.3/lib/node_modules/firebase-tools/lib/emulator/functionsEmulatorRuntime.js:451:9
    at Generator.next (<anonymous>)
    at /Users/Jascha/.nvm/versions/node/v10.15.3/lib/node_modules/firebase-tools/lib/emulator/functionsEmulatorRuntime.js:7:71
    at new Promise (<anonymous>)
    at __awaiter (/Users/Jascha/.nvm/versions/node/v10.15.3/lib/node_modules/firebase-tools/lib/emulator/functionsEmulatorRuntime.js:3:12)
    at main (/Users/Jascha/.nvm/versions/node/v10.15.3/lib/node_modules/firebase-tools/lib/emulator/functionsEmulatorRuntime.js:421:12)
    at Object.<anonymous> (/Users/Jascha/.nvm/versions/node/v10.15.3/lib/node_modules/firebase-tools/lib/emulator/functionsEmulatorRuntime.js:511:5)
```

If I downgrade back to firebase-tools `6.8.0` it works.

**Comment by tranvietthanh1997** [comment]
> @diogolessa Downgrading to firebase-tools 6.8.0 also fixes the problem, is that not an option for you?

Thank you, it's worked for me
